### PR TITLE
DB version checkup

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Verify if the database schema matches the current version of packetfence
## Impacts

Verify if the database schema matches the current version of packetfence
## NEWS file entries

New Features
++++++++++++
- Verify if the database schema matches the current version of packetfence
## Delete branch after merge

NO

### CODE REVIEW

File: lib/pf/constants.pm [hunk 1]
```diff
@@ -16,8 +16,10 @@ use warnings;
 use Readonly;
 use base qw(Exporter);
 our @EXPORT = qw(
-    $FALSE $TRUE $YES $NO $default_pid $admin_pid
+    $FALSE $TRUE $YES $NO $default_pid $admin_pid $PF_VERSION
 );
+
+Readonly::Scalar our $PF_VERSION => '5.0.0';
```
Comment (julsemaan) [inline lib/pf/constants.pm]: I would use the content of pf-release rather than this

Comment (jrouzierinverse) [inline lib/pf/constants.pm]: Done

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 2]
```diff
@@ -99,3 +99,16 @@ UPDATE `node` SET pid = 'default' WHERE status = 'unreg' AND pid = 'admin';
 ALTER TABLE node ADD dhcp_vendor VARCHAR(255) AFTER dhcp_fingerprint,
                  ADD device_type VARCHAR(255) AFTER dhcp_vendor,
                  ADD device_class VARCHAR(255) AFTER device_type;
+
```
Comment (dwlfrth) [inline db/upgrade-X.X.X-X.Y.Z.sql]: You are missing a -- line

File: db/pf-schema-X.Y.Z.sql [hunk 3]
```diff
@@ -902,3 +902,16 @@ CREATE TABLE keyed (
   value LONGBLOB,
   PRIMARY KEY(id)
 ) ENGINE=InnoDB;
+--
+-- Table structure for table 'pf_version'
+--
+
+CREATE TABLE pf_version (
+    `version` varchar(15) NOT NULL
```
Comment (dwlfrth) [inline db/pf-schema-X.Y.Z.sql]: varchar(15) ? that's a huge release number ;)

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 4]
```diff
@@ -99,3 +99,16 @@ UPDATE `node` SET pid = 'default' WHERE status = 'unreg' AND pid = 'admin';
...
+-- Table structure for table 'pf_version'
+--
+
+CREATE TABLE pf_version (
+    `version` varchar(15) NOT NULL
```
Comment (dwlfrth) [inline db/upgrade-X.X.X-X.Y.Z.sql]: varchar(15) ? that's a huge release number ;)

Comment (dwlfrth) [inline db/upgrade-X.X.X-X.Y.Z.sql]: s/varchar/VARCHAR

Comment (dwlfrth) [inline db/upgrade-X.X.X-X.Y.Z.sql]: It can also be on one single line

File: db/pf-schema-X.Y.Z.sql (hunk 3 above)
Comment (dwlfrth) [inline db/pf-schema-X.Y.Z.sql]: s/varchar/VARCHAR

Comment (dwlfrth) [inline db/pf-schema-X.Y.Z.sql]: It can also be on one single line

File: db/pf-schema-X.Y.Z.sql [hunk 5]
```diff
@@ -902,3 +902,16 @@ CREATE TABLE keyed (
   value LONGBLOB,
   PRIMARY KEY(id)
 ) ENGINE=InnoDB;
+--
```
Comment (dwlfrth) [inline db/pf-schema-X.Y.Z.sql]: I'd skip a line before 905

File: lib/pf/pfcmd/checkup.pm [hunk 6]
```diff
@@ -1082,6 +1084,19 @@ sub apache_filter_rules {
...
+=item db_version
+
+Make sure the database matches the current version
+Return the current version if it matches or undef if it does not match
+
+=cut
+
+sub db_version {
+    unless(pf::version::version_check()) {
+        add_problem ( $FATAL, "the database does not match the current version $PF_VERSION" );
```
Comment (dwlfrth) [inline lib/pf/pfcmd/checkup.pm]: Should it really be 'FATAL' ? I mean... PF _can_ work but may run into some issues...
I'm asking a question here

File: lib/pf/pfcmd/checkup.pm [hunk 7]
```diff
@@ -1082,6 +1084,19 @@ sub apache_filter_rules {
     }
 }
 
+=item db_version
+
+Make sure the database matches the current version
+Return the current version if it matches or undef if it does not match
```
Comment (dwlfrth) [inline lib/pf/pfcmd/checkup.pm]: Is that line accurate ?

File: lib/pf/version.pm [hunk 8]
```diff
@@ -0,0 +1,120 @@
+package pf::version;
+=head1 NAME
+
+pf::version add documentation
```
Comment (dwlfrth) [inline lib/pf/version.pm]: Then, add it ! :-P

File: lib/pf/version.pm [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    return 1;
+}
+
+=head2 version_check
+
+Checks the version of db
+
+=cut
+
+sub version_check {
```
Comment (dwlfrth) [inline lib/pf/version.pm]: We are in pf::version class. I'm just wondering if we could simply rename version_check to check... ?
Something like what I've done in pf::iplog...

File: lib/pf/version.pm [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+   return $row->[0];
+}
+
+=head2 version_release
+
+Get the current release of packetence
+
+=cut
+
+sub version_release {
```
Comment (dwlfrth) [inline lib/pf/version.pm]: Shouldn't it be "get_pf_release_version"

File: lib/pf/pfcmd/checkup.pm (hunk 6 above)
Comment (rouzier) [inline lib/pf/pfcmd/checkup.pm]: This does not block the httpd.admin from starting.
That is the only guarantee we can make in my opinion.

**Comment by jrouzierinverse** [comment]
Please re-review

**Comment by louismunro** [comment]
Please add a way for us to track every version that the database schema has ever been.
This would allow us, among other things, to check and see if an update has been skipped between two versions.

**Comment by julsemaan** [comment]
+1 on Louis' comment

**Comment by jrouzierinverse** [comment]
Please re-review

File: db/pf_version.sql [hunk 11]
```diff
@@ -0,0 +1,15 @@
+
+SET @MAJOR_VERSION = 5,@MINOR_VERSION = 1, @PATCH_LEVEL = 0;
+SET @VERSION_INT = @MAJOR_VERSION << 16 | @MINOR_VERSION << 8 | @PATCH_LEVEL;
```
Comment (dwlfrth) [inline db/pf_version.sql]: Drop me

File: db/pf_version.sql [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+-- Table structure for table 'pf_version'
+--
+
+CREATE TABLE pf_version ( `id` INT NOT NULL, `version` VARCHAR(11) NOT NULL , PRIMARY KEY (id));
+
+--
+-- Updating to current version
+--
+
+INSERT INTO pf_version (id,version) VALUES (@VERSION_INT,CONCAT(@MAJOR_VERSION,'.',@MINOR_VERSION,'.',@PATCH_LEVEL));
```
Comment (dwlfrth) [inline db/pf_version.sql]: I'm there too ;)

File: db/pf_version.sql [hunk 13]
```diff
@@ -0,0 +1,15 @@
+
+SET @MAJOR_VERSION = 5,@MINOR_VERSION = 1, @PATCH_LEVEL = 0;
+SET @VERSION_INT = @MAJOR_VERSION << 16 | @MINOR_VERSION << 8 | @PATCH_LEVEL;
+
+--
+-- Table structure for table 'pf_version'
+--
+
+CREATE TABLE pf_version ( `id` INT NOT NULL, `version` VARCHAR(11) NOT NULL , PRIMARY KEY (id));
```
Comment (dwlfrth) [inline db/pf_version.sql]: Rework me based on VERSION_INT removal

File: lib/pf/version.pm [hunk 14]
```diff
@@ -0,0 +1,119 @@
+package pf::version;
+=head1 NAME
+
+pf::version
+
+=cut
+
+=head1 DESCRIPTION
+
+pf::version
```
Comment (dwlfrth) [inline lib/pf/version.pm]: Description != name

**Comment by dwlfrth** [comment]
db/pf-version.sql ... 
- Shouldn't be part of db/pf-schema...
- How to "run" it otherwise

**Comment by dwlfrth** [comment]
Also need to take care of UPGRADE path...

File: lib/pf/version.pm [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+sub version_db_prepare {
+
+    $version_statements->{'version_check_sql'} = get_db_handle()->prepare(qq[
+            SELECT version FROM pf_version WHERE version = ?;
+    ]);
+    $version_db_prepared = 1;
+    return 1;
+}
+
+=head2 version_check
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_check/db_schema_version_check

File: lib/pf/version.pm [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+Checks the version of db
```
Comment (dwlfrth) [inline lib/pf/version.pm:54]: Check the version of the database schema according to the current version of the PacketFence release.

File: lib/pf/version.pm [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+=cut
+
+sub version_check {
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_check/db_schema_version_check

File: lib/pf/version.pm [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+=head2 version_db_prepare
+
+Initialize database prepared statements
+
+=cut
+
+sub version_db_prepare {
+
+    $version_statements->{'version_check_sql'} = get_db_handle()->prepare(qq[
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_check_sql/db_schema_version_check_sql

File: lib/pf/version.pm [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+=head2 version_check
+
+Checks the version of db
+
+=cut
+
+sub version_check {
+   my $sth = db_query_execute(PF_VERSION, $version_statements, 'version_check_sql', version_current());
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_check_sql/db_schema_version_check_sql
s/version_current/get_current

File: lib/pf/version.pm [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+   return unless $sth;
+   my $row = $sth->fetch;
+   $sth->finish;
+   return unless $row;
+   return $row->[0];
+}
+
+=head2 version_release
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_release/get_packetfence_release

File: lib/pf/version.pm [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+Get the current release of packetence
```
Comment (dwlfrth) [inline lib/pf/version.pm:84]: s/packetfence/PacketFence

File: lib/pf/version.pm [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+=cut
+
+sub version_release {
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_release/get_packetfence_release

File: lib/pf/version.pm [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    my ( $pfrelease_fh, $release );
+    open( $pfrelease_fh, '<', "$conf_dir/pf-release" )
+        || get_logger->logdie("Unable to open $conf_dir/pf-release: $!");
+    $release = <$pfrelease_fh>;
+    close($pfrelease_fh);
+    chomp($release);
+    return $release ;
+}
+
+=head2 version_current
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_current/get_current

File: lib/pf/version.pm [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+Get the current version of packetence
```
Comment (dwlfrth) [inline lib/pf/version.pm:100]: Get the current version of the PacketFence release

File: lib/pf/version.pm [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+
+=cut
+
+sub version_current {
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_current/get_current

File: lib/pf/version.pm [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+    my $release = version_release();
```
Comment (dwlfrth) [inline lib/pf/version.pm]: s/version_release/get_packetfence_release

**Comment by dwlfrth** [comment]
A lot of my comments are implicating name changing but, from my point of view, code should be readable easily. Especially for something simple like this.
Having to check the code to see what version_XYZ and version_ZYX returns can be avoided by simply modifying the naming.

**Comment by dwlfrth** [comment]
get_packetfence_release could also be get_release to keep the 'get_version' way...

**Comment by jrouzierinverse** [comment]
Our current naming convention for functions in most of our db modules as you know is modulename_[something].

So in order to keep our naming convention consistent.

I will do the following

version_check   => version_check_db

version_current => version_get_current

version_release => version_get_release

**Comment by dwlfrth** [comment]
I know but if I remember correctly, that was only to match sql statements that needs to be prefixed by the module name...
I don't really see the point of having a function called version_get_current in the pf::version module...
pf::version::version_get_current...

I tried to move out of this "standard" with the refactor of pf::iplog...

**Comment by dwlfrth** [comment]
Anyway... prefix it or not, at least as it is "readable"

File: db/pf_version.sql (hunk 13 above)
Comment (jrouzierinverse) [inline db/pf_version.sql]: When we spoke about the VERSION_INT I did not remember why it was done (I know a comment would have helped).
It was done to ensure we can properly order the version when we have double or triple digit major/minor/patch numbers.

If we depend on the string value alone then we will have the following order for the version numbers.

5.1.0
5.10.0
5.2.0
5.3.0
5.4.0
5.5.0
5.6.0
5.7.0
5.8.0
5.9.0

The VERSION_INT is constructed in such a way where the version can be sorted properly. 
(As long as no major/minor/patch number goes above 255).

When updating the new version number in the DB you just change the appropriate version variable.

Comment (dwlfrth) [inline db/pf_version.sql]: That's an awesome comment to put in :dancer:

Comment (dwlfrth) [inline db/pf_version.sql]: Hmm... forget about the "dancer"... meant to be :D

File: db/pf_version.sql (hunk 11 above)
Comment (dwlfrth) [inline db/pf_version.sql]: We'll keep it

File: db/pf_version.sql (hunk 12 above)
Comment (dwlfrth) [inline db/pf_version.sql]: Keeping it

**Comment by jrouzierinverse** [comment]
Please Review

**Comment by dwlfrth** [comment]
This is ready but I think we should wait before merging it.
- Wait until the very last moment before 5.1.0 (at least a minor version and not a sub-minor/patch version)
- Wait for "standardized" release process based on branches discussions. Will need to talk about it.

**Comment by jrouzierinverse** [comment]
Delaying merging this pull request will only make it harder to find any last minute bugs before releasing 5.1.0. 

(Of course I have no bugs in my code https://www.youtube.com/watch?v=nnwWKkNau4I )

@dwlfrth What is your concern about this feature that make you want to delay merging it.

We should either move it to milestone 5.2.0 or merge it now so we can see if this will break builds

**Comment by dwlfrth** [comment]
We need to discuss the way to handle branches merge while doing major, minor and sub-minor releases.
We'll wait until that.

**Comment by louismunro** [comment]
Not sure I want us to wait until the last minute for something that will affect all upgrades.
It should be tested thoroughly by then. 
Let's fix the branch handling rather than delay important features and testing.
The process is not what we are releasing.

**Comment by dwlfrth** [comment]
As I said, we need to clear out how we manage merges to branches when it comes to major, minor, sub-minor releases.
That will affect the versioning behavior and the different 'schema' files.
We will talk about it tomorrow when @fdurand will have time.
Until then, I do not see any urge in merging it since it is a pretty simple thing that will probably not require THAT MUCH of testing.

**Comment by dwlfrth** [comment]
Merged...
Please have a quick look at https://github.com/inverse-inc/packetfence/commit/213c38e802a831ac7992fa5027e7556cd2e32aed

**Comment by dwlfrth** [comment]
@jrouzierinverse 
Also, please advise on line 23 of pf::version...

**Pull request merged**